### PR TITLE
Feature/tie filters together

### PR DESCRIPTION
**Pull request by fpdotmonkey**
This ties the peep, pip, and tidal volume calculations to the back end.  You can see the results on the front end by running this code.

```bash
./setup.sh
. venv/bin/activate
docker build -t zmq_proxy:latest .
docker run --rm -p 8000:8000 zmq_proxy:latest
```

You can then open http://localhost:8000 in a browser and see all the data changing.  Note that it'll look funny in Firefox because they don't support WebKit, but the stuff that's relevant to this PR will be visible.  Also all patients will have the same data.

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -25,39 +25,44 @@ def add_datum(self, datum):
...
 def main():
     sensors = Sensors()
-    sensor_data = sensors.poll()
     calculator = Calculator()
     communicator = Communicator()
-    running = True
-    while running:
-        try:
+    try:
+        while True:
```
Comment (cporter) [inline server.py:53]: Why did `running` go away?

File: server.py [hunk 2]
```diff
@@ -25,39 +25,44 @@ def add_datum(self, datum):
...
             start_time = time.time()
-            calculator.add_datum(sensor_data)
-
-            communicator.publish_message(calculator.get_datum())
+            data = sensors.poll()
+            print(data)
```
Comment (cporter) [inline server.py:56]: Check out the `logging` module for this kind of printf debugging stuff. You can keep it in your production code but disable shouty mode at runtime.

File: tidal_volume.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class TidalVolume():
+
+    def __init__(self, initial_time):
+        self._integral_filter = CausalIntegralFilter(0, initial_time)
+
+    def append(self, flow_rate, t):
+        self._integral_filter.append(flow_rate, t)
+        if math.isclose(flow_rate, 0, abs_tol=0.01):
```
Comment (cporter) [inline tidal_volume.py:13]: Is `0.01` all the more epsilon you'll need? Might be worth at least declaring a constant such that future readers have a little more information about why it's a particular value.

File: server.py [hunk 4]
```diff
@@ -25,39 +25,44 @@ def add_datum(self, datum):
...
 
 class Communicator():
 
     def __init__(self, port=5000):
-        self._socket = zmq.Context().socket(zmq.PUB)
+        self._context = zmq.Context()
+        self._socket = self._context.socket(zmq.PUB)
         self._socket.bind(f"tcp://*:{port}")
 
+    def close(self):
```
Comment (cporter) [inline server.py:39]: What happens if `close` is called twice?

File: process_sample_data.py [hunk 5]
```diff
@@ -84,16 +85,31 @@ def _parseData(self):
...
 
             if (pressure_marker in datum):
                 self._pressures.append(float(splitDatum[3]
                                              .replace(pressure_marker,
                                                       "")
                                              .strip("\n")) / 10)
+
+    def _remove_repeat_data(self):
+        non_repeat_indices = []
+        for i in range(len(self._timestamps)):
```
Comment (cporter) [inline process_sample_data.py:102]: Can you populate `non_repeat_indices` as a comprehension as opposed to an append inside a loop?

File: process_sample_data.py [hunk 6]
```diff
@@ -84,16 +85,31 @@ def _parseData(self):
...
+            if (not (math.isclose(self._flow_rates[i],
+                                  self._flow_rates[i-1])
+                     and math.isclose(self._tidal_volumes[i],
+                                      self._tidal_volumes[i-1])
+                     and (math.isclose(self._pressures[i],
+                                       self._pressures[i-1])
+                          if len(self._pressures) > 0 else True))):
```
Comment (cporter) [inline process_sample_data.py:109]: Is the length of `_pressures` ever different from the length of the other lists?

File: process_sample_data.py [hunk 7]
```diff
@@ -20,9 +21,9 @@ class ProcessSampleData():
     """
 
     def __init__(self, path_to_data):
-        self._flow_data_file = open(path_to_data, "r")
-        self._parseData()
-        self._flow_data_file.close()
+        with open(path_to_data, "r") as self._flow_data_file:
+            self._parse_data()
+            self._remove_repeat_data()
```
Comment (cporter) [inline process_sample_data.py:26]: I gather that this is for testing, so efficiency isn't a huge concern, but you could save yourself some time if you dedup while you parse. You could reformulate the inners of `_parse_data` as a stream of tuples, keep track of previous and only append to the data lists if the new values are meaningfully different.

**Review by cporter — COMMENTED** [review]
I have a few notes for code clarity and one efficiency suggestion. Looks reasonable overall. Why doesn't Firefox work?

File: process_sample_data.py (hunk 6 above)
Comment (fpdotmonkey) [reply process_sample_data.py:109]: yes, I have one data set that has flow rate, tidal volume, and pressure, and another than only has flow rate and tidal volume.

File: server.py (hunk 1 above)
Comment (fpdotmonkey) [reply server.py:53]: well there was never an instance where `running` could be False, and I'm intending on my next PR to be something that fixes this.  I suppose I'll revert till the next PR.

**Comment by fpdotmonkey** [comment]
> Why doesn't Firefox work?

I rely on a WebKit feature (drawing polygons or something) that Firefox doesn't support.  Not a big deal since it only has to work on 1 platform.

File: process_sample_data.py (hunk 6 above)
Comment (cporter) [reply process_sample_data.py]: Ok, so what's the desired behavior out past the end of the shorter of the two? It's not clear at this point.

File: process_sample_data.py [hunk 8]
```diff
@@ -67,33 +67,41 @@ def _parseData(self):
         tidal_volume_marker = "TidalVol:"
         pressure_marker = "Pressurex10:"
 
+        previous_data = [None, None, math.inf]
```
Comment (cporter) [inline process_sample_data.py:70]: Is this the only place that `previous_data` is ever set? Looks like you're missing a reset down in the previous_data equality check branch.

File: server.py (hunk 4 above)
Comment (cporter) [reply server.py:39]: Ping.

File: process_sample_data.py [hunk 9]
```diff
@@ -67,33 +67,41 @@ def _parseData(self):
         tidal_volume_marker = "TidalVol:"
         pressure_marker = "Pressurex10:"
```
Comment (cporter) [inline process_sample_data.py:68]: Nitpick, and perhaps I'm ignorant of this project's coding style guidelines, but constants are often expressed as `ALL_CAPS`, especially when there's no type system to signify their purpose.

File: process_sample_data.py [hunk 10]
```diff
@@ -20,9 +21,8 @@ class ProcessSampleData():
     """
 
     def __init__(self, path_to_data):
-        self._flow_data_file = open(path_to_data, "r")
-        self._parseData()
-        self._flow_data_file.close()
+        with open(path_to_data, "r") as self._flow_data_file:
```
Comment (cporter) [inline process_sample_data.py:24]: This could be an argument rather than a class member. That's a little more clear, and a little more testable.

**Review by cporter — CHANGES_REQUESTED** [review]
The `close()` question is still open (ha!), I think you missed setting a variable in the data filtering -- might need a test for that -- and a quick style issue re: constants.

File: server.py (hunk 4 above)
Comment (fpdotmonkey) [reply server.py:39]: I commented on this before, but it seems like it didn't show up for you.  Nothing happens if you close twice.  It's just closed.

**Comment by fpdotmonkey** [comment]
> What happens if `close` is called twice?

Nothing.  The stream is just closed, no exceptions.

File: server.py (hunk 4 above)
Comment (cporter) [reply server.py:39]: Fair enough. (And polite of the zmq library to provide that behavior.) A common pattern in situations like this is to make sure you only close a resource once, just to guard against aberrant behaviors from unpolite libraries.

**Review by cporter — APPROVED** [review]

**Pull request merged**